File: backend/accounts/guidelines.py
"""
Guideline validation helpers.

A Guideline is a named set of listing requirements for a specific building
managed by a ManagementCompany. Each field is optional — null means no
requirement for that field.
"""

VALID_FURNISHED_STATUSES = ("furnished", "unfurnished", "partially_furnished")


def validate_guideline_data(data: dict) -> tuple[bool, str]:
    """
    Validate incoming guideline field data (used for create/update).
    Returns (is_valid, error_message).
    """
    if not isinstance(data, dict):
        return False, "Guideline data must be an object."

    name = data.get("name", "").strip()
    if not name:
        return False, "Guideline name is required."

    min_rent = data.get("min_rent")
    max_rent = data.get("max_rent")
    if min_rent is not None and (not isinstance(min_rent, (int, float)) or min_rent < 0):
        return False, "min_rent must be a non-negative number."
    if max_rent is not None and (not isinstance(max_rent, (int, float)) or max_rent < 0):
        return False, "max_rent must be a non-negative number."
    if min_rent is not None and max_rent is not None and min_rent > max_rent:
        return False, "min_rent cannot be greater than max_rent."

    min_deposit = data.get("min_deposit")
    max_deposit = data.get("max_deposit")
    if min_deposit is not None and (not isinstance(min_deposit, (int, float)) or min_deposit < 0):
        return False, "min_deposit must be a non-negative number."
    if max_deposit is not None and (not isinstance(max_deposit, (int, float)) or max_deposit < 0):
        return False, "max_deposit must be a non-negative number."
    if min_deposit is not None and max_deposit is not None and min_deposit > max_deposit:
        return False, "min_deposit cannot be greater than max_deposit."

    min_days = data.get("min_availability_days")
    if min_days is not None and (not isinstance(min_days, int) or min_days < 1):
        return False, "min_availability_days must be a positive integer."

    furnished_status = data.get("furnished_status")
    if furnished_status is not None and furnished_status not in VALID_FURNISHED_STATUSES:
        return False, f"furnished_status must be one of: {', '.join(VALID_FURNISHED_STATUSES)}."

    required_amenities = data.get("required_amenities")
    if required_amenities is not None:
        if not isinstance(required_amenities, list):
            return False, "required_amenities must be a list."
        if not all(isinstance(a, str) for a in required_amenities):
            return False, "Each amenity in required_amenities must be a string code."

    return True, ""


def guideline_to_human(guideline) -> list[str]:
    """
    Convert a Guideline model instance to a list of human-readable requirement strings.
    Only includes fields that have a value set.
    """
    lines = []

    if guideline.min_rent is not None and guideline.max_rent is not None:
        lines.append(f"Monthly rent between ${guideline.min_rent:,.0f} and ${guideline.max_rent:,.0f}")
    elif guideline.min_rent is not None:
        lines.append(f"Monthly rent at least ${guideline.min_rent:,.0f}")
    elif guideline.max_rent is not None:
        lines.append(f"Monthly rent at most ${guideline.max_rent:,.0f}")

    if guideline.min_deposit is not None and guideline.max_deposit is not None:
        lines.append(f"Security deposit between ${guideline.min_deposit:,.0f} and ${guideline.max_deposit:,.0f}")
    elif guideline.min_deposit is not None:
        lines.append(f"Security deposit at least ${guideline.min_deposit:,.0f}")
    elif guideline.max_deposit is not None:
        lines.append(f"Security deposit at most ${guideline.max_deposit:,.0f}")

    if guideline.min_availability_days is not None:
        lines.append(f"Available for at least {guideline.min_availability_days} days")

    if guideline.utilities_included is True:
        lines.append("Utilities must be included")
    elif guideline.utilities_included is False:
        lines.append("Utilities must not be included")

    if guideline.pets_allowed is True:
        lines.append("Must allow pets")
    elif guideline.pets_allowed is False:
        lines.append("Must not allow pets")

    if guideline.furnished_status:
        lines.append(f"Must be {guideline.furnished_status.replace('_', ' ')}")

    for code in (guideline.required_amenities or []):
        lines.append(f"Must include amenity: {code}")

    return lines

File: frontend/src/types/guidelines.ts
export type GuidelineType =
  | "rent_range"
  | "deposit_range"
  | "min_availability_days"
  | "utilities_included"
  | "pets_allowed"
  | "furnished_status"
  | "amenity_required";

export type FurnishedStatus = "furnished" | "unfurnished" | "partially_furnished";

export interface RentRangeGuideline {
  type: "rent_range";
  min_rent?: number;
  max_rent?: number;
}

export interface DepositRangeGuideline {
  type: "deposit_range";
  min_deposit?: number;
  max_deposit?: number;
}

export interface MinAvailabilityDaysGuideline {
  type: "min_availability_days";
  min_days: number;
}

export interface UtilitiesIncludedGuideline {
  type: "utilities_included";
  required: boolean;
}

export interface PetsAllowedGuideline {
  type: "pets_allowed";
  required: boolean;
}

export interface FurnishedStatusGuideline {
  type: "furnished_status";
  value: FurnishedStatus;
}

export interface AmenityRequiredGuideline {
  type: "amenity_required";
  amenity_code: string;
  amenity_label: string;
}

export type Guideline =
  | RentRangeGuideline
  | DepositRangeGuideline
  | MinAvailabilityDaysGuideline
  | UtilitiesIncludedGuideline
  | PetsAllowedGuideline
  | FurnishedStatusGuideline
  | AmenityRequiredGuideline;

export const GUIDELINE_TYPE_LABELS: Record<GuidelineType, string> = {
  rent_range: "Monthly Rent Range",
  deposit_range: "Security Deposit Range",
  min_availability_days: "Minimum Availability",
  utilities_included: "Utilities Included",
  pets_allowed: "Pets Allowed",
  furnished_status: "Furnished Status",
  amenity_required: "Required Amenity",
};

export function guidelineToHuman(g: Guideline): string {
  switch (g.type) {
    case "rent_range": {
      const { min_rent, max_rent } = g;
      if (min_rent != null && max_rent != null)
        return `Monthly rent between $${min_rent.toLocaleString()} and $${max_rent.toLocaleString()}`;
      if (min_rent != null) return `Monthly rent at least $${min_rent.toLocaleString()}`;
      if (max_rent != null) return `Monthly rent at most $${max_rent.toLocaleString()}`;
      return "Monthly rent range";
    }
    case "deposit_range": {
      const { min_deposit, max_deposit } = g;
      if (min_deposit != null && max_deposit != null)
        return `Security deposit between $${min_deposit.toLocaleString()} and $${max_deposit.toLocaleString()}`;
      if (min_deposit != null) return `Security deposit at least $${min_deposit.toLocaleString()}`;
      if (max_deposit != null) return `Security deposit at most $${max_deposit.toLocaleString()}`;
      return "Security deposit range";
    }
    case "min_availability_days":
      return `Available for at least ${g.min_days} day${g.min_days === 1 ? "" : "s"}`;
    case "utilities_included":
      return g.required ? "Utilities must be included" : "Utilities must not be included";
    case "pets_allowed":
      return g.required ? "Must allow pets" : "Must not allow pets";
    case "furnished_status":
      return `Must be ${g.value.replace(/_/g, " ")}`;
    case "amenity_required":
      return `Must include: ${g.amenity_label}`;
  }
}
